doc/src/tutorial/matrices.rst

```diff
@@ -231,6 +231,149 @@ filled with `0`\ s.
     ⎢           ⎥
     ⎣0   0  0  5⎦
 
+
+Random Matrices
+---------------
+
+To create Matrices with specific properties use one of
+the random matrix generation functions, e.g.
+``invertible``, ``triangular``, ``orthogonal``.
+Find a :ref:`full list of random matrices <matrices-random>`.
+
+    >>> from sympy.matrices.random import rand
+    >>> rand.seed(1)
+
+    >>> from sympy.matrices.random import invertible, triangular, orthogonal
+    >>> from sympy import simplify
+
+    >>> M = invertible(3)  # only integer entries
+    >>> M
+    ⎡-1  -1  -1⎤
+    ⎢          ⎥
+    ⎢-2  -1  -1⎥
+    ⎢          ⎥
+    ⎣0   0   1 ⎦
+    >>> M.inv()
+    ⎡1   -1  0 ⎤
+    ⎢          ⎥
+    ⎢-2  1   -1⎥
+    ⎢          ⎥
+    ⎣0   0   1 ⎦
+
+    >>> M = invertible(3, scalar_set=(sqrt(2),))  # entries as sums and products of sqrt(2) and 1
```

Comment:
I'm not sure about `scalar_set`. I think perhaps a domain argument could be good. Ideally there should be some way for the user to pass in some distribution of expressions that they have designed themselves.

Reply:
I previously wrote some code for random matrix generation that I didn't finish although my emphasis was more on generating different kinds of matrix expressions and in particular working with domain elements. I wonder if an approach like that can be used here:
```python
from pytest import raises

from random import randint, randrange, random, sample, Random
from itertools import product

from sympy import QQ, ZZ, GF, ZZ_I, QQ_I, sqrt
from sympy.abc import x, y
from sympy.polys.matrices import DomainMatrix



class RandomDomain:
    """Superclass for random element generators"""
    rng = Random()

    def __call__(self, rng=None):
        return self.generate(rng or self.rng)


class RandomGF(RandomDomain):
    """Random finite field element GF(p)"""
    def __init__(self, K):
        self.domain = K

    def generate(self, rng):
        return self.domain(rng.randint(0, self.domain.mod))


class _RandomZZ(RandomDomain):
    """Base class for RandomZZ and RandomZZpos"""

    domain = ZZ

    def __init__(self, bits):
        self.bits = bits
        self.intmax = 1 << bits - 1


class RandomZZ(_RandomZZ):
    """Uniform random integer with maximum bitsize"""
    def generate(self, rng):
        return ZZ(rng.randint(-self.intmax, self.intmax))


class RandomZZpos(_RandomZZ):
    """Uniform random positive integer ZZ with maximum bitsize"""
    def generate(self, rng):
        return ZZ(rng.randint(1, self.intmax))


class RandomZZlog(_RandomZZ):
    """Random positive integer ZZ the logarithmic distribution"""
    def generate(self, rng):
        logbits = 0
        bits = self.bits
        while bits:
            bits //= 2
            logbits += 1
        bits = rng.randint(1, 1 << logbits - 1)
        return ZZ(rng.randint(0, 1 << bits - 1))


class RandomQQ(RandomDomain):
    """Random rational QQ with given num/den distribution"""

    domain = QQ

    def __init__(self, numrand, denrand):
        self.numrand = numrand
        self.denrand = denrand

    @classmethod
    def bits(cls, bits):
        """RandomQQ with uniform num/den distribution and max bits"""
        return RandomQQ(RandomZZ(bits), RandomZZpos(bits))

    def generate(self, rng):
        return QQ(self.numrand(rng), self.denrand(rng))


class _RandomGaussian(RandomDomain):
    """Base class for RandomZZ_I and RandomQQ_I"""
    def __init__(self, rand_x, rand_y=None):
        rand_y = rand_y or rand_x
        self.rand_x = rand_x
        self.rand_y = rand_y

    def generate(self, rng):
        return self.domain(self.rand_x(rng), self.rand_y(rng))


class RandomZZ_I(_RandomGaussian):
    """Random Gaussian integer ZZ_I with given x,y distribution"""
    domain = ZZ_I


class RandomQQ_I(_RandomGaussian):
    """Random Gaussian rational QQ_I with given x,y distribution"""
    domain = QQ_I


class RandomALG(RandomDomain):
    """Random element of an algebraic number field QQ(a)"""
    def __init__(self, K, rand_coeff):
        self.domain = K
        self.degree = K.mod.degree()
        self.rand = rand_coeff

    def generate(self, rng):
        coeffs = [self.rand(rng) for _ in range(self.degree + 1)]
        return self.domain.new(coeffs)


class RandomPolyRing(RandomDomain):
    """Random polynomial in K[x,y]"""
    def __init__(self, R, rand_coeff, degree):
        self.domain = R
        self.rand = rand_coeff
        self.gens = R.gens
        self.degree = degree
        self.monomials = list(product(range(degree+1), repeat=len(R.gens)))

    def generate(self, rng):
        gens = self.gens
        poly_d = {m: self.rand(rng) for m in self.monomials}
        return self.domain.new(poly_d)


class RandomFracField(RandomDomain):
    """Random element of a fraction field K(x,y)"""
    def __init__(self, K, rand_poly_num, rand_poly_den=None):
        self.domain = K
        self.rand_num = rand_poly_num
        self.rand_den = rand_poly_den or rand_poly_num

    def generate(self, rng):
        num = self.rand_num(rng)
        den = self.rand_den(rng)
        return self.domain(num) / self.domain(den)


class RandomDDM(RandomDomain):
    """Random dense DomainMatrix"""
    def __init__(self, rand_element, shape):
        self.rand = rand_element
        self.shape = shape
        self.domain = rand_element.domain

    def generate(self, rng):
        rand = self.rand
        rows, cols = self.shape
        elements = [[rand(rng) for _ in range(cols)] for _ in range(rows)]
        dM = DomainMatrix(elements, self.shape, self.domain)
        return dM


class RandomSDM(RandomDomain):
    """Random sparse DomainMatrix"""
    def __init__(self, rand_element, shape, density):
        self.rand = rand_element
        self.shape = shape
        self.density = density
        self.domain = rand_element.domain

    def generate(self, rng):
        rand = self.rand
        rows, cols = self.shape
        num_nonzero = round((rows * cols) * self.density)
        indices = rng.sample(range(rows * cols), num_nonzero)
        sdm = {}
        for ij in indices:
            i, j = divmod(ij, cols)
            val = rand(rng)
            try:
                sdm[i][j] = val
            except KeyError:
                sdm[i] = {j: val}
        dM = DomainMatrix(sdm, self.shape, self.domain)
        return dM


def random_domain_generator(domain, **kwargs):
    """Make a random generator for the given domain"""
    bits = kwargs.get('bits', 6)
    degree = kwargs.get('degree', 2)
    logarithmic = kwargs.get('logarithmic', False)

    if domain.is_FiniteField:
        return RandomGF(domain)

    elif domain == ZZ:
        if not logarithmic:
            return RandomZZ(bits)
        else:
            return RandomZZlog(bits)

    elif domain == QQ:
        return RandomQQ.bits(bits)

    elif domain == ZZ_I:
        return RandomZZ_I(RandomZZ(bits))

    elif domain == QQ_I:
        return RandomQQ_I(RandomQQ.bits(bits))

    elif domain.is_AlgebraicField:
        return RandomALG(domain, RandomQQ.bits(bits))

    elif domain.is_PolynomialRing:
        rand_coeff = random_domain_generator(domain.domain, **kwargs)
        return RandomPolyRing(domain, rand_coeff, degree)

    elif domain.is_FractionField:
        rand_poly = random_domain_generator(domain.get_ring(), **kwargs)
        return RandomFracField(domain, rand_poly)

    else:
        raise ValueError


def random_element(domain, rng=None, **kwargs):
    """Return a random element from the given domain"""
    rand = random_domain_generator(domain, **kwargs)
    return rand(rng)


def random_matrix(domain_or_rand, shape, *, rng=None, density=1, **kwargs):
    """Random matrix with elements of given domain"""
    rows, cols = shape

    if isinstance(domain_or_rand, RandomDomain):
        rand = domain_or_rand
    else:
        domain = domain_or_rand
        rand = random_domain_generator(domain, **kwargs)

    if density == 1:
        rand_DM = RandomDDM(rand, shape)
    else:
        rand_DM = RandomSDM(rand, shape, density)

    dM = rand_DM(rng)

    return dM


def test_rand_element():
    seed = Random().randint(1, 10**6)
    rng1 = Random(seed)
    rng2 = Random(seed)

    domains = [
        ZZ, QQ, GF(7), ZZ_I, QQ_I,
        QQ.algebraic_field(sqrt(2)),
        ZZ[x], QQ[x],
        ZZ[x,y], QQ[x,y],
        ZZ.frac_field(x), QQ.frac_field(x),
        ZZ.frac_field(x,y), QQ.frac_field(x,y),
    ]
    for K in domains:
        e = random_element(K)
        assert K.of_type(e)

        assert random_element(K, rng=rng1) == random_element(K, rng=rng2)

        for density in [0.5, 1]:
            for r in [K, random_domain_generator(K)]:
                dM = random_matrix(r, (3, 3), density=density)
                assert isinstance(dM, DomainMatrix)
                assert dM.domain == K

                dM1 = random_matrix(r, (3, 3), density=density, rng=rng1)
                dM2 = random_matrix(r, (3, 3), density=density, rng=rng2)
                assert dM1 == dM2

    from sympy.polys.domains import Domain
    class Undefined(Domain):
        def __init__(self):
            pass
    K = Undefined()
    raises(ValueError, lambda: random_domain_generator(K))
```

Reply:
The main point is that with someone can design their own random expression generator and pass it in to `random_matrix` to get a matrix generated from that distribution of expressions.

Reply:
I think, I get the idea. Note, there are mainly two different ways a rng is used. 

1. for sampling _dimensions_/_shapes_, _indices_ or _sign_ in `sgn * sqrt(1 - c**2)` which are probably uniform distributed 
2. for sampling matrix entries. But here, one has to distinguish between 
   - _isometries_ which take elements of the circle of complex units `scalar_set` as an argument
   - _elementary_  which takes elements `scalar_set`  **and** units `unit_set` of a ring as arguments
   - _normal forms_ which take eigenvalues `spec` as arguments

My suggestion would be, let `scalar_set`,  `unit_set` and `spec` possibly be a rng, too. If so, they should implement a `sample(k)` method. Otherwise (e.g. if just a `list` or `tuple`) they will be sampled using the _default rng_ `rand.sample()` which is given by an extra _seed_ argument.

Here what it could look like:
```python
def diagonal_normal(dim, spec=None, seed=None):
    """ diagonal normal form"""
    
    rng = seed if has_attr(seed, 'sample') else random.Random(seed)

    if has_attr(spec, 'sample'):
        spec = tuple(spec.sample(1) for _ in range(dim))
    elif not len(spec)==dim:
        spec = tuple(rng.sample(spec, 1) for _ in range(dim))
```

Then the matrix entries can come from some `RandomDomain` or just a finite `list`. The `list` is very handy to draw easy problems for undergraduate _linear algebra_ courses. Note, even a few elements could give a quite huge matrices. Arbitrary rationals would give very easy unsolvable problems. 

```
>>> orthogonal(3, scalar_set=(0, 1/2, sqrt(2)/2, sqrt(3)/2))
⎡   ⎛  1   √2⎞      ⎛√2   1⎞             ⎛√2   1⎞      ⎛  1   √2⎞⎤
⎢√2⋅⎜- ─ + ──⎟   √2⋅⎜── + ─⎟          √2⋅⎜── + ─⎟   √2⋅⎜- ─ + ──⎟⎥
⎢   ⎝  2   4 ⎠      ⎝4    2⎠             ⎝4    2⎠      ⎝  2   4 ⎠⎥
⎢───────────── + ───────────  1/2   - ─────────── + ─────────────⎥
⎢      2              2                    2              2      ⎥
⎢                                                                ⎥
⎢   ⎛  1   √2⎞      ⎛√2   1⎞             ⎛  1   √2⎞      ⎛√2   1⎞⎥
⎢√2⋅⎜- ─ + ──⎟   √2⋅⎜── + ─⎟          √2⋅⎜- ─ + ──⎟   √2⋅⎜── + ─⎟⎥
⎢   ⎝  2   4 ⎠      ⎝4    2⎠             ⎝  2   4 ⎠      ⎝4    2⎠⎥
⎢───────────── + ───────────  1/2   - ───────────── + ───────────⎥
⎢      2              2                     2              2     ⎥
⎢                                                                ⎥
⎢            √2               -√2                                ⎥
⎢            ──               ────                0              ⎥
⎣            2                 2                                 ⎦

>>> orthogonal(3, length=10)
⎡                 √2   1               1   √2      ⎤
⎢  -1/2           ── + ─               ─ - ──      ⎥
⎢                 4    2               2   4       ⎥
⎢                                                  ⎥
⎢                    ⎛  1   √2⎞     ⎛  1   √2⎞     ⎥
⎢                 √2⋅⎜- ─ - ──⎟  √2⋅⎜- ─ - ──⎟     ⎥
⎢ 1   √2     √2      ⎝  2   4 ⎠     ⎝  2   4 ⎠   √2⎥
⎢ ─ - ──   - ── - ─────────────  ───────────── - ──⎥
⎢ 2   4      4          2              2         4 ⎥
⎢                                                  ⎥
⎢                     ⎛1   √2⎞             ⎛1   √2⎞⎥
⎢                  √2⋅⎜─ - ──⎟          √2⋅⎜─ - ──⎟⎥
⎢  1   √2     √2      ⎝2   4 ⎠     √2      ⎝2   4 ⎠⎥
⎢- ─ - ──   - ── - ───────────   - ── + ───────────⎥
⎣  2   4      4         2          4         2     ⎦
```

Reply:
btw, no problem to rename `scalar_set` and `unit_set` arguments by `domain` and `units`. 

As mentioned above, it might be more consistent to use `units` instead of `scalar_set` for isometries like `orthogonal` or `unitary` (as `cos/sin` are just representing a point on the _circle of complex units_).

Reply:
I'm not sure what is a good interface for seeding here. What you suggest sounds reasonable.

Reply:
Me either. I've been thinking more than a year.. 

Nevertheless latest version offers at least most flexibility - by the cost of heavily overloading the `seed` argument. Hope it is intuitive enough. 

To explain it, I added a extra section in [`doc/src/modules/matrices/random.rst`](https://github.com/jan-philipp-hoffmann/sympy/blob/pre_master/doc/src/modules/matrices/random.rst).